lib/logmatcher.c

```diff
@@ -384,8 +424,26 @@ typedef struct _LogMatcherPcreMatchResult
   const gchar *source_value;
   gssize source_value_len;
   pcre2_match_data *match_data;
+  gboolean force_direct_set;
 } LogMatcherPcreMatchResult;
 
+static void
+log_matcher_pcre_match_result_init(LogMatcherPcreMatchResult *self, LogMatcherPcreRe *matcher,
+                                   const gchar *source_value, gssize source_value_len, NVHandle source_handle)
+{
+  self->match_data = pcre2_match_data_create_from_pattern(matcher->pattern, NULL);
+  self->source_value = source_value;
+  self->source_value_len = source_value_len != -1 ? source_value_len : strlen(source_value);
+  self->source_handle = source_handle;
+  self->force_direct_set = _is_input_overwritten_by_named_capture(matcher, source_handle);
```

Comment:
I think this function does not need to be called from every match operation. It's enough to call it just once, during initialization. Doing it for every match seems to be a tad expensive to me.

Reply:
The problem is that the `value_handle` only comes with `log_matcher_match()` or `log_matcher_match_value()`. Compile time we do not know which value will be used for matching, or a not a value at all (buffer, template).

I have checked the code, and we always call `match()` or `match_value()` with the same value handle on the call sites, so I can store the value handle in init, but it seems strange to me.

Maybe we could cache the last matched value handle and its overwrite behavior, and only recalculate the overwrite behavior if the cached value handle is different than the currently matched one.

What do you think?

Reply:
we've discussed this IRL and concluded that we should go back to the initial solution, e.g. we would stop producing indirect nv-pairs as soon as we discover that we are overwriting the original value.

Those that are already saved as indirect matches will be copied automatically by NVTable.